Drop merging of authorized_keys.d into authorized_keys

bgilbert opened the pull request:
Fixes #157.  Container Linux will need to build with `--features cl-legacy`.

bgilbert committed:
providers: move AuthorizedKeyEntry out of providers

Wrap keys for update_ssh_keys directly in write_ssh_keys().

bgilbert committed:
providers: move update_ssh_keys wrapper to helper function

bgilbert committed:
src: gate update_ssh_keys behind "cl-legacy" feature

It should only be enabled on Container Linux, for compatibility with
existing deployments.

bgilbert committed:
docs: document distro integration with --ssh-keys option

bgilbert committed:
travis: also test in cl-legacy mode

lucab commented on src/providers/mod.rs:
```diff
@@ -34,11 +34,19 @@ pub mod vagrant_virtualbox;
+#[cfg(not(feature = "legacy"))]
+use users::os::unix::UserExt;
+#[cfg(not(feature = "legacy"))]
+use tempfile;
```
I think you can plainly drop this.

lucab replied:
Discussed out-of-band: `tempfile::foo` is supported in newer compiler, but our baseline toolchain doesn't have that yet. Let's keep this.

lucab commented on src/providers/mod.rs:
```diff
@@ -34,11 +34,19 @@ pub mod vagrant_virtualbox;
 
 use std::collections::HashMap;
 use std::fs::{self, File};
+#[cfg(not(feature = "legacy"))]
```
If you want to avoid all these conditionals below, you can move most of the `use` statements directly inside function body.

lucab commented on Cargo.toml:
```diff
@@ -35,9 +35,13 @@ slog-async = "2.1"
 users = "^0.8.1"
 
+[features]
+legacy = ["update-ssh-keys"]
```
Features are usually additive. Do we want to call this something more specific like `merge-authorized-keys`? If we are going to add more CL-legacy features, we can always add an umbrella one called `cl-legacy`. Thoughts?

lucab replied:
Discussed out-of-band: this will likely ends up entangled in other things, such as project renaming. Let's stick to a monolithic `cl-legacy` to keep complexity low.

lucab reviewed (commented):
Code looks fine already. I only have a couple nits and a doubt on feature naming.

bgilbert committed:
providers: implement non-cl-legacy path for SSH key writing

Don't use update_ssh_keys and don't merge authorized_keys.d fragments
into authorized_keys.  Instead, just write the fragment file safely,
using as the tempfile a dotfile in the same directory.  If there are no
keys, delete the fragment file.

lucab reviewed (approved):
LGTM!
